File: src/preview_tool/artifact_git.py
"""Local-only Git history for host-owned Preview artifacts."""

from __future__ import annotations

import os
import stat
import subprocess
import tempfile
from pathlib import Path

GITIGNORE = """/previews/.locks/
/previews/.partial/
/previews/.previous/
/previews/.records/.*
/previews/.preview-exchange-probe-*/
/.preview-exchange-probe-*/
/.gitignore.*
/.in-progress-plugin.partial/
/.in-progress-plugin.previous/
"""
GIT_TIMEOUT_SECONDS = 30


def _environment() -> dict[str, str]:
    return {
        "GIT_CONFIG_GLOBAL": "/dev/null",
        "GIT_CONFIG_NOSYSTEM": "1",
        "GIT_NO_REPLACE_OBJECTS": "1",
        "GIT_OPTIONAL_LOCKS": "0",
        "GIT_PAGER": "cat",
        "GIT_TERMINAL_PROMPT": "0",
        "LANG": "C.UTF-8",
        "LC_ALL": "C.UTF-8",
        "PAGER": "cat",
        "PATH": "/usr/bin:/bin",
    }


def _git(
    root: Path,
    arguments: list[str],
    *,
    allowed: frozenset[int] = frozenset({0}),
) -> str:
    completed = subprocess.run(
        [
            "git",
            "-c",
            "core.autocrlf=false",
            "-c",
            "core.fsmonitor=false",
            "-c",
            "core.hooksPath=/dev/null",
            "-c",
            "commit.gpgSign=false",
            *arguments,
        ],
        cwd=root,
        env=_environment(),
        stdin=subprocess.DEVNULL,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        timeout=GIT_TIMEOUT_SECONDS,
        check=False,
    )
    if completed.returncode not in allowed:
        detail = (completed.stderr or completed.stdout).strip()[-2_000:]
        raise RuntimeError(f"artifact Git command failed: {detail}")
    return completed.stdout.strip()


def _ensure_gitignore(root: Path) -> None:
    path = root / ".gitignore"
    try:
        metadata = path.lstat()
    except FileNotFoundError:
        metadata = None
    if metadata is not None and (
        not stat.S_ISREG(metadata.st_mode) or stat.S_ISLNK(metadata.st_mode)
    ):
        raise RuntimeError(f"artifact .gitignore is not a regular file: {path}")
    if metadata is not None and path.read_text(encoding="utf-8") == GITIGNORE:
        return
    descriptor, temporary_name = tempfile.mkstemp(prefix=".gitignore.", dir=root)
    temporary = Path(temporary_name)
    try:
        with os.fdopen(descriptor, "w", encoding="utf-8") as handle:
            handle.write(GITIGNORE)
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(temporary, path)
    finally:
        try:
            temporary.unlink()
        except FileNotFoundError:
            pass


def prepare_artifact_git(root: Path) -> Path:
    """Initialize and validate the exact local-only artifact repository."""
    resolved = root.resolve(strict=True)
    metadata = resolved.lstat()
    if not stat.S_ISDIR(metadata.st_mode) or stat.S_ISLNK(metadata.st_mode):
        raise RuntimeError(f"artifact root is not a real directory: {resolved}")
    git_directory = resolved / ".git"
    try:
        git_metadata = git_directory.lstat()
    except FileNotFoundError:
        git_metadata = None
    if git_metadata is None:
        _git(resolved, ["init", "--initial-branch=main"])
    elif not stat.S_ISDIR(git_metadata.st_mode) or stat.S_ISLNK(git_metadata.st_mode):
        raise RuntimeError(f"artifact Git metadata is not a real directory: {git_directory}")

    repository_root = Path(_git(resolved, ["rev-parse", "--show-toplevel"])).resolve()
    if repository_root != resolved:
        raise RuntimeError(f"artifact Git root does not match artifact root: {repository_root}")
    if _git(resolved, ["symbolic-ref", "--quiet", "--short", "HEAD"]) != "main":
        raise RuntimeError("artifact Git repository must remain on branch main")
    if _git(resolved, ["remote"]):
        raise RuntimeError("artifact Git repository must not configure a remote")

    staged = _git(resolved, ["diff", "--cached", "--name-only", "-z"])
    if any(
        path != ".gitignore"
        and not path.startswith("previews/")
        and not path.startswith("in-progress-plugin/")
        for path in staged.split("\0")
        if path
    ):
        raise RuntimeError("artifact Git index contains changes outside Preview-owned paths")

    _ensure_gitignore(resolved)
    return resolved


def snapshot_artifacts(root: Path) -> str | None:
    """Stage bounded artifact paths and commit changed validated bytes."""
    resolved = prepare_artifact_git(root)
    paths = [".gitignore", "previews", "in-progress-plugin"]
    _git(resolved, ["add", "--all", "--", *paths])
    _git(resolved, ["diff", "--cached", "--check", "--no-ext-diff"])
    status = subprocess.run(
        [
            "git",
            "-c",
            "core.autocrlf=false",
            "diff",
            "--cached",
            "--quiet",
            "--no-ext-diff",
        ],
        cwd=resolved,
        env=_environment(),
        stdin=subprocess.DEVNULL,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.PIPE,
        timeout=GIT_TIMEOUT_SECONDS,
        check=False,
    )
    if status.returncode == 0:
        _git(resolved, ["rev-parse", "HEAD"])
        _verify_snapshot(resolved)
        return None
    if status.returncode != 1:
        detail = status.stderr.decode("utf-8", errors="replace").strip()[-2_000:]
        raise RuntimeError(f"artifact Git status failed: {detail}")
    _git(
        resolved,
        [
            "-c",
            "user.name=Preview",
            "-c",
            "user.email=preview@localhost",
            "commit",
            "-m",
            "preview: validated artifacts → local snapshot",
        ],
    )
    revision = _git(resolved, ["rev-parse", "HEAD"])
    _verify_snapshot(resolved)
    return revision


def _verify_snapshot(root: Path) -> None:
    if _git(root, ["remote"]):
        raise RuntimeError("artifact Git repository gained a remote during snapshot")
    dirty = _git(
        root,
        [
            "status",
            "--porcelain=v1",
            "--untracked-files=normal",
            "--",
            ".gitignore",
            "previews",
            "in-progress-plugin",
        ],
    )
    if dirty:
        raise RuntimeError("artifact files changed during the local Git snapshot")


__all__ = ["GITIGNORE", "prepare_artifact_git", "snapshot_artifacts"]

File: src/artifact-git.ts
import {
  closeSync,
  constants,
  existsSync,
  fsyncSync,
  lstatSync,
  openSync,
  readFileSync,
  realpathSync,
  renameSync,
  unlinkSync,
  writeFileSync,
} from "node:fs"
import { join } from "node:path"
import { randomUUID } from "node:crypto"
import { spawnSync } from "node:child_process"

export const GITIGNORE = `/previews/.locks/
/previews/.partial/
/previews/.previous/
/previews/.records/.*
/previews/.preview-exchange-probe-*/
/.preview-exchange-probe-*/
/.gitignore.*
/.in-progress-plugin.partial/
/.in-progress-plugin.previous/
`

const gitEnvironment = {
  GIT_CONFIG_GLOBAL: "/dev/null",
  GIT_CONFIG_NOSYSTEM: "1",
  GIT_NO_REPLACE_OBJECTS: "1",
  GIT_OPTIONAL_LOCKS: "0",
  GIT_PAGER: "cat",
  GIT_TERMINAL_PROMPT: "0",
  LANG: "C.UTF-8",
  LC_ALL: "C.UTF-8",
  PAGER: "cat",
  PATH: "/usr/bin:/bin",
}

function git(root: string, arguments_: readonly string[], allowed = new Set([0])): string {
  const result = spawnSync(
    "git",
    [
      "-c",
      "core.autocrlf=false",
      "-c",
      "core.fsmonitor=false",
      "-c",
      "core.hooksPath=/dev/null",
      "-c",
      "commit.gpgSign=false",
      ...arguments_,
    ],
    {
      cwd: root,
      encoding: "utf8",
      env: gitEnvironment,
      input: "",
      timeout: 30_000,
    },
  )
  if (result.error || result.signal || !allowed.has(result.status ?? -1)) {
    const detail = (result.stderr || result.stdout || result.error?.message || "Git failed")
      .trim()
      .slice(-2_000)
    throw new Error(`artifact Git command failed: ${detail}`)
  }
  return result.stdout.trim()
}

function gitStatus(root: string, arguments_: readonly string[]): number {
  const result = spawnSync(
    "git",
    [
      "-c",
      "core.autocrlf=false",
      "-c",
      "core.fsmonitor=false",
      "-c",
      "core.hooksPath=/dev/null",
      "-c",
      "commit.gpgSign=false",
      ...arguments_,
    ],
    {
      cwd: root,
      env: gitEnvironment,
      input: "",
      timeout: 30_000,
    },
  )
  if (result.error || result.signal || result.status === null) {
    throw new Error("artifact Git status failed")
  }
  return result.status
}

function ensureGitignore(root: string): void {
  const path = join(root, ".gitignore")
  if (existsSync(path)) {
    const metadata = lstatSync(path)
    if (!metadata.isFile() || metadata.isSymbolicLink()) {
      throw new Error(`artifact .gitignore is not a regular file: ${path}`)
    }
    if (readFileSync(path, "utf8") === GITIGNORE) return
  }
  const temporary = join(root, `.gitignore.${randomUUID()}`)
  const descriptor = openSync(
    temporary,
    constants.O_WRONLY | constants.O_CREAT | constants.O_EXCL,
    0o600,
  )
  try {
    writeFileSync(descriptor, GITIGNORE, "utf8")
    fsyncSync(descriptor)
    closeSync(descriptor)
    renameSync(temporary, path)
  } catch (error) {
    try {
      closeSync(descriptor)
    } catch {
      // The descriptor was already closed.
    }
    throw error
  } finally {
    try {
      unlinkSync(temporary)
    } catch {
      // Rename or earlier cleanup removed the temporary file.
    }
  }
}

export function prepareArtifactGit(root: string): string {
  const resolved = realpathSync(root)
  const metadata = lstatSync(resolved)
  if (!metadata.isDirectory() || metadata.isSymbolicLink()) {
    throw new Error(`artifact root is not a real directory: ${resolved}`)
  }
  const gitDirectory = join(resolved, ".git")
  if (!existsSync(gitDirectory)) git(resolved, ["init", "--initial-branch=main"])
  else {
    const gitMetadata = lstatSync(gitDirectory)
    if (!gitMetadata.isDirectory() || gitMetadata.isSymbolicLink()) {
      throw new Error(`artifact Git metadata is not a real directory: ${gitDirectory}`)
    }
  }
  if (realpathSync(git(resolved, ["rev-parse", "--show-toplevel"])) !== resolved) {
    throw new Error("artifact Git root does not match artifact root")
  }
  if (git(resolved, ["symbolic-ref", "--quiet", "--short", "HEAD"]) !== "main") {
    throw new Error("artifact Git repository must remain on branch main")
  }
  if (git(resolved, ["remote"]))
    throw new Error("artifact Git repository must not configure a remote")
  const staged = git(resolved, ["diff", "--cached", "--name-only", "-z"])
  if (
    staged
      .split("\0")
      .filter(Boolean)
      .some(
        (path) =>
          path !== ".gitignore" &&
          !path.startsWith("previews/") &&
          !path.startsWith("in-progress-plugin/"),
      )
  ) {
    throw new Error("artifact Git index contains changes outside Preview-owned paths")
  }
  ensureGitignore(resolved)
  return resolved
}

function verifySnapshot(root: string): void {
  if (git(root, ["remote"]))
    throw new Error("artifact Git repository gained a remote during snapshot")
  const dirty = git(root, [
    "status",
    "--porcelain=v1",
    "--untracked-files=normal",
    "--",
    ".gitignore",
    "previews",
    "in-progress-plugin",
  ])
  if (dirty) throw new Error("artifact files changed during the local Git snapshot")
}

export function snapshotArtifacts(root: string): string | undefined {
  const resolved = prepareArtifactGit(root)
  git(resolved, ["add", "--all", "--", ".gitignore", "previews", "in-progress-plugin"])
  git(resolved, ["diff", "--cached", "--check", "--no-ext-diff"])
  const status = gitStatus(resolved, ["diff", "--cached", "--quiet", "--no-ext-diff"])
  if (status === 0) {
    verifySnapshot(resolved)
    return undefined
  }
  if (status !== 1) throw new Error("artifact Git status failed")
  git(resolved, [
    "-c",
    "user.name=Preview",
    "-c",
    "user.email=preview@localhost",
    "commit",
    "-m",
    "preview: validated artifacts → local snapshot",
  ])
  const revision = git(resolved, ["rev-parse", "HEAD"])
  verifySnapshot(resolved)
  return revision
}
